File: src/dataHealth/appDataIngressPipeline.ts
import type { AppData } from '../models/training-model';
import type { AppDataRepairRegistry } from './appDataRepairEngine';
import { getAppDataRepairRegistry } from './appDataRepairRegistry';
import type {
  DataHealthAutoRepairSummary,
  RepairTrigger,
} from './appDataRepairTypes';
import {
  type AutoRepairOrchestratorResult,
  readAutoRepairSummary,
  runAutoRepairOrchestrator,
} from './autoRepairOrchestrator';
import type { AutoRepairBackupAdapter } from './autoRepairBackupAdapter';
import { buildCleanAppDataView, type CleanAppDataView } from './cleanAppDataView';
import {
  evaluateCloudUploadEligibility,
  type CloudUploadEligibility,
} from './uploadEligibility';
import { computeAppDataHash } from './repairs/repairHelpers';

export type AppDataIngressSource =
  | 'boot'
  | 'localStorage-load'
  | 'import-restore'
  | 'backup-restore'
  | 'cloud-restore'
  | 'cloud-pull'
  | 'read-mirror'
  | 'cloud-parity'
  | 'account-switch'
  | 'post-session-complete'
  | 'pre-training-decision'
  | 'pre-cloud-upload'
  | 'export';

export type UploadEligibilityMode = 'check' | 'enforce' | 'ignore';

export type PassiveStatusTone = 'ok' | 'auto-repaired' | 'audit-pending' | 'backup-failed' | 'busy';

export interface AppDataIngressInput {
  source: AppDataIngressSource;
  appData: AppData;
  accountId?: string;
  allowMutation?: boolean;
  allowAutoRepair?: boolean;
  requireBackup?: boolean;
  uploadEligibilityMode?: UploadEligibilityMode;
  registry?: AppDataRepairRegistry;
  backupAdapter?: AutoRepairBackupAdapter;
  now?: () => Date;
  operationId?: string;
}

export interface AppDataIngressResult {
  source: AppDataIngressSource;
  operationId: string;
  cleanView: CleanAppDataView;
  repairedAppData?: AppData;
  repairSummary?: DataHealthAutoRepairSummary;
  orchestratorResult?: AutoRepairOrchestratorResult;
  shouldPersist: boolean;
  shouldBlockCloudUpload: boolean;
  uploadEligibility: CloudUploadEligibility;
  passiveStatus: { line: string; tone: PassiveStatusTone };
  warnings: string[];
  triggeredOrchestrator: boolean;
  appDataHashBefore: string;
  appDataHashAfter: string;
}

interface SourceDefaults {
  allowMutation: boolean;
  allowAutoRepair: boolean;
  requireBackup: boolean;
  uploadEligibilityMode: UploadEligibilityMode;
  repairTrigger: RepairTrigger;
}

const SOURCE_DEFAULTS: Record<AppDataIngressSource, SourceDefaults> = {
  'boot':                   { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'check',   repairTrigger: 'boot' },
  'localStorage-load':      { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'check',   repairTrigger: 'boot' },
  'import-restore':         { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'check',   repairTrigger: 'import' },
  'backup-restore':         { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'check',   repairTrigger: 'import' },
  'cloud-restore':          { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'enforce', repairTrigger: 'cloud_restore' },
  'cloud-pull':             { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'check',   repairTrigger: 'cloud_restore' },
  'read-mirror':            { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'check',   repairTrigger: 'audit' },
  'cloud-parity':           { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'check',   repairTrigger: 'audit' },
  'account-switch':         { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'enforce', repairTrigger: 'boot' },
  'post-session-complete':  { allowMutation: true,  allowAutoRepair: true,  requireBackup: true,  uploadEligibilityMode: 'check',   repairTrigger: 'post_session' },
  'pre-training-decision':  { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'ignore',  repairTrigger: 'audit' },
  'pre-cloud-upload':       { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'enforce', repairTrigger: 'audit' },
  'export':                 { allowMutation: false, allowAutoRepair: false, requireBackup: false, uploadEligibilityMode: 'ignore',  repairTrigger: 'manual' },
};

const FORBIDDEN_AUTO_REPAIR_WITHOUT_MUTATION: AppDataIngressSource[] = [
  'cloud-pull',
  'read-mirror',
  'cloud-parity',
  'pre-training-decision',
  'pre-cloud-upload',
  'export',
];

const composePassiveStatus = (params: {
  source: AppDataIngressSource;
  appliedCount: number;
  pendingRepairs: number;
  auditOnly: number;
  backupFailed: boolean;
}): { line: string; tone: PassiveStatusTone } => {
  if (params.backupFailed) {
    return { line: '数据正在自动整理，稍后同步', tone: 'backup-failed' };
  }
  if (params.appliedCount > 0) {
    return { line: `已自动修复 ${params.appliedCount} 个旧版本问题`, tone: 'auto-repaired' };
  }
  if (params.pendingRepairs > 0) {
    return { line: `${params.pendingRepairs} 个待自动修复`, tone: 'busy' };
  }
  if (params.auditOnly > 0) {
    return { line: `${params.auditOnly} 个已隔离，不影响训练建议`, tone: 'audit-pending' };
  }
  return { line: '数据已自动检查', tone: 'ok' };
};

const generateOperationId = (source: AppDataIngressSource, accountId: string | undefined, hash: string): string => {
  const scope = accountId ?? 'unscoped';
  const random = Math.floor(Math.random() * 1_000_000).toString(36);
  return `ingress_${source}_${scope}_${hash.slice(-8)}_${Date.now().toString(36)}_${random}`;
};

export const processIncomingAppData = async (
  input: AppDataIngressInput,
): Promise<AppDataIngressResult> => {
  const defaults = SOURCE_DEFAULTS[input.source];
  if (!defaults) {
    throw new Error(`[dataHealth] unknown ingress source: ${input.source}`);
  }
  const allowMutation = input.allowMutation ?? defaults.allowMutation;
  const allowAutoRepair = input.allowAutoRepair ?? defaults.allowAutoRepair;
  const requireBackup = input.requireBackup ?? defaults.requireBackup;
  const uploadEligibilityMode = input.uploadEligibilityMode ?? defaults.uploadEligibilityMode;

  if (allowAutoRepair && FORBIDDEN_AUTO_REPAIR_WITHOUT_MUTATION.includes(input.source) && !input.allowMutation) {
    throw new Error(`[dataHealth] ingress source ${input.source} cannot allowAutoRepair without explicit allowMutation=true`);
  }

  const now = input.now || (() => new Date());
  const registry = input.registry || getAppDataRepairRegistry();
  const appDataHashBefore = computeAppDataHash(input.appData);
  const operationId = input.operationId || generateOperationId(input.source, input.accountId, appDataHashBefore);

  const cleanView = buildCleanAppDataView(input.appData, { now });

  let orchestratorResult: AutoRepairOrchestratorResult | undefined;
  let repairedAppData: AppData | undefined;
  let workingAppData: AppData = input.appData;
  let triggeredOrchestrator = false;
  const warnings: string[] = [];

  if (allowMutation && allowAutoRepair) {
    try {
      orchestratorResult = await runAutoRepairOrchestrator({
        appData: input.appData,
        triggeredBy: defaults.repairTrigger,
        registry,
        backupAdapter: input.backupAdapter,
        now,
      });
      triggeredOrchestrator = true;
      if (orchestratorResult.changed) {
        repairedAppData = orchestratorResult.appData;
        workingAppData = orchestratorResult.appData;
      } else {
        workingAppData = orchestratorResult.appData;
      }
      orchestratorResult.warnings.forEach((warning) => warnings.push(warning));
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      warnings.push(`orchestrator_error: ${message}`);
    }
  }

  const uploadEligibility = evaluateCloudUploadEligibility(workingAppData, { registry, now });

  let shouldBlockCloudUpload = false;
  if (uploadEligibilityMode === 'enforce') {
    shouldBlockCloudUpload = !uploadEligibility.eligible;
  } else if (uploadEligibilityMode === 'check') {
    shouldBlockCloudUpload = !uploadEligibility.eligible;
  } else {
    shouldBlockCloudUpload = false;
  }

  const auditFindingsCount = orchestratorResult?.auditFindings.length ?? uploadEligibility.auditOnly;
  const appliedCount = orchestratorResult?.results.filter((entry) => entry.status === 'applied').length ?? 0;

  const passiveStatus = composePassiveStatus({
    source: input.source,
    appliedCount,
    pendingRepairs: uploadEligibility.pendingRepairs,
    auditOnly: auditFindingsCount,
    backupFailed: uploadEligibility.backupFailed,
  });

  const shouldPersist = Boolean(allowMutation && repairedAppData);

  if (requireBackup && allowMutation && allowAutoRepair && orchestratorResult && !orchestratorResult.backup && orchestratorResult.changed === false && uploadEligibility.backupFailed) {
    warnings.push('backup_failed: ingress kept runtime guard active without mutation');
  }

  const summary = readAutoRepairSummary(workingAppData);

  return {
    source: input.source,
    operationId,
    cleanView,
    repairedAppData,
    repairSummary: summary,
    orchestratorResult,
    shouldPersist,
    shouldBlockCloudUpload,
    uploadEligibility,
    passiveStatus,
    warnings,
    triggeredOrchestrator,
    appDataHashBefore,
    appDataHashAfter: computeAppDataHash(workingAppData),
  };
};

export const ingressSourceDefaults = (source: AppDataIngressSource): SourceDefaults => SOURCE_DEFAULTS[source];

File: ios/packages/IronPathDataHealth/Sources/IronPathDataHealth/AppDataIngressPipeline.swift
// AppDataIngressPipeline — iOS-3C ingress pipeline.
//
// Swift port of `src/dataHealth/appDataIngressPipeline.ts`. Wraps
// the iOS-3A CleanAppDataView + iOS-3B AutoRepairOrchestrator into a
// single entry-point that boot / import / cloud-restore / etc. flows
// can call with `AppData` + `AppDataIngressSource`.
//
// iOS-3C explicitly DOES NOT port the cloud-upload-eligibility check
// from `src/dataHealth/uploadEligibility.ts` — that lands in iOS-7
// when CloudKit / Supabase wiring arrives. Until then, the result
// struct exposes only the runtime-guard + repair surface.

import Foundation
import IronPathDomain

// MARK: - Source enum

public enum AppDataIngressSource: String, Equatable, Hashable, Sendable, CaseIterable {
    case boot
    case localStorageLoad   = "localStorage-load"
    case importRestore      = "import-restore"
    case backupRestore      = "backup-restore"
    case cloudRestore       = "cloud-restore"
    case cloudPull          = "cloud-pull"
    case readMirror         = "read-mirror"
    case cloudParity        = "cloud-parity"
    case accountSwitch      = "account-switch"
    case postSessionComplete = "post-session-complete"
    case preTrainingDecision = "pre-training-decision"
    case preCloudUpload     = "pre-cloud-upload"
    case export
}

public enum AppDataIngressPassiveStatusTone: String, Equatable, Hashable, Sendable {
    case ok
    case autoRepaired = "auto-repaired"
    case auditPending = "audit-pending"
    case backupFailed = "backup-failed"
    case busy
}

public struct AppDataIngressPassiveStatus: Equatable, Sendable {
    public let line: String
    public let tone: AppDataIngressPassiveStatusTone

    public init(line: String, tone: AppDataIngressPassiveStatusTone) {
        self.line = line
        self.tone = tone
    }
}

// MARK: - Per-source defaults

public struct AppDataIngressSourceDefaults: Equatable, Sendable {
    public let allowMutation: Bool
    public let allowAutoRepair: Bool
    public let requireBackup: Bool
    public let repairTrigger: RepairTrigger

    public init(
        allowMutation: Bool,
        allowAutoRepair: Bool,
        requireBackup: Bool,
        repairTrigger: RepairTrigger
    ) {
        self.allowMutation = allowMutation
        self.allowAutoRepair = allowAutoRepair
        self.requireBackup = requireBackup
        self.repairTrigger = repairTrigger
    }
}

public func ingressSourceDefaults(_ source: AppDataIngressSource) -> AppDataIngressSourceDefaults {
    switch source {
    case .boot:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .boot)
    case .localStorageLoad:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .boot)
    case .importRestore:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .importing)
    case .backupRestore:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .importing)
    case .cloudRestore:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .cloudRestore)
    case .cloudPull:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .cloudRestore)
    case .readMirror:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .audit)
    case .cloudParity:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .audit)
    case .accountSwitch:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .boot)
    case .postSessionComplete:
        return .init(allowMutation: true, allowAutoRepair: true, requireBackup: true, repairTrigger: .postSession)
    case .preTrainingDecision:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .audit)
    case .preCloudUpload:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .audit)
    case .export:
        return .init(allowMutation: false, allowAutoRepair: false, requireBackup: false, repairTrigger: .manual)
    }
}

private let forbiddenAutoRepairWithoutMutation: Set<AppDataIngressSource> = [
    .cloudPull,
    .readMirror,
    .cloudParity,
    .preTrainingDecision,
    .preCloudUpload,
    .export,
]

// MARK: - Input / Result

public struct AppDataIngressOptions: Sendable {
    public let accountId: String?
    public let allowMutation: Bool?
    public let allowAutoRepair: Bool?
    public let requireBackup: Bool?
    public let backupAdapter: AutoRepairBackupAdapter?
    public let registry: RepairRegistry?
    public let operationId: String?

    public init(
        accountId: String? = nil,
        allowMutation: Bool? = nil,
        allowAutoRepair: Bool? = nil,
        requireBackup: Bool? = nil,
        backupAdapter: AutoRepairBackupAdapter? = nil,
        registry: RepairRegistry? = nil,
        operationId: String? = nil
    ) {
        self.accountId = accountId
        self.allowMutation = allowMutation
        self.allowAutoRepair = allowAutoRepair
        self.requireBackup = requireBackup
        self.backupAdapter = backupAdapter
        self.registry = registry
        self.operationId = operationId
    }
}

public struct AppDataIngressResult: Sendable {
    public let source: AppDataIngressSource
    public let operationId: String
    public let cleanView: CleanAppDataView
    /// Non-nil only when the orchestrator actually mutated AppData.
    public let repairedAppData: AppData?
    public let repairSummary: DataHealthAutoRepairSummary?
    public let orchestratorResult: AutoRepairOrchestratorResult?
    public let shouldPersist: Bool
    public let passiveStatus: AppDataIngressPassiveStatus
    public let warnings: [String]
    public let triggeredOrchestrator: Bool
    public let appDataHashBefore: String
    public let appDataHashAfter: String

    public init(
        source: AppDataIngressSource,
        operationId: String,
        cleanView: CleanAppDataView,
        repairedAppData: AppData?,
        repairSummary: DataHealthAutoRepairSummary?,
        orchestratorResult: AutoRepairOrchestratorResult?,
        shouldPersist: Bool,
        passiveStatus: AppDataIngressPassiveStatus,
        warnings: [String],
        triggeredOrchestrator: Bool,
        appDataHashBefore: String,
        appDataHashAfter: String
    ) {
        self.source = source
        self.operationId = operationId
        self.cleanView = cleanView
        self.repairedAppData = repairedAppData
        self.repairSummary = repairSummary
        self.orchestratorResult = orchestratorResult
        self.shouldPersist = shouldPersist
        self.passiveStatus = passiveStatus
        self.warnings = warnings
        self.triggeredOrchestrator = triggeredOrchestrator
        self.appDataHashBefore = appDataHashBefore
        self.appDataHashAfter = appDataHashAfter
    }
}

public enum AppDataIngressError: Error, Equatable, Sendable {
    case forbiddenAutoRepair(AppDataIngressSource)
}

// MARK: - Entry point

public func processIncomingAppData(
    appData: AppData,
    source: AppDataIngressSource,
    clock: RuntimeGuardClock = SystemRuntimeGuardClock(),
    options: AppDataIngressOptions = AppDataIngressOptions()
) throws -> AppDataIngressResult {
    let defaults = ingressSourceDefaults(source)
    let allowMutation = options.allowMutation ?? defaults.allowMutation
    let allowAutoRepair = options.allowAutoRepair ?? defaults.allowAutoRepair
    let requireBackup = options.requireBackup ?? defaults.requireBackup

    // Read-only sources cannot opt into auto-repair without also
    // explicitly opting into mutation.
    if allowAutoRepair && forbiddenAutoRepairWithoutMutation.contains(source) && options.allowMutation != true {
        throw AppDataIngressError.forbiddenAutoRepair(source)
    }

    let registry = options.registry ?? fullRepairRegistry()
    let appDataHashBefore = computeAppDataHash(appData)
    let operationId = options.operationId
        ?? generateOperationId(source: source, accountId: options.accountId, hash: appDataHashBefore, clock: clock)

    // Always build CleanAppDataView, even on read-only sources.
    let cleanView = buildCleanAppDataView(appData, clock: clock)

    var orchestratorResult: AutoRepairOrchestratorResult?
    var repairedAppData: AppData?
    var workingAppData = appData
    var triggeredOrchestrator = false
    var warnings: [String] = []

    if allowMutation && allowAutoRepair {
        do {
            let result = try runAutoRepairOrchestrator(AutoRepairOrchestratorInput(
                appData: appData,
                triggeredBy: defaults.repairTrigger,
                registry: registry,
                backupAdapter: options.backupAdapter,
                clock: clock
            ))
            orchestratorResult = result
            triggeredOrchestrator = true
            if result.changed {
                repairedAppData = result.appData
                workingAppData = result.appData
            } else {
                workingAppData = result.appData
            }
            for warning in result.warnings {
                warnings.append(warning)
            }
        } catch {
            warnings.append("orchestrator_error: \(error)")
        }
    }

    let summary = readAutoRepairSummary(workingAppData)
    let auditFindingsCount = orchestratorResult?.auditFindings.count ?? 0
    let appliedCount = orchestratorResult?.results.filter { $0.status == .applied }.count ?? 0
    let backupFailed = orchestratorResult?.warnings.contains(where: { $0.contains("backup_failed") }) ?? false

    let passiveStatus = composePassiveStatus(
        appliedCount: appliedCount,
        pendingRepairs: 0,  // iOS-3C does NOT port uploadEligibility; iOS-7 will own pending counts
        auditOnly: auditFindingsCount,
        backupFailed: backupFailed
    )

    if requireBackup && allowMutation && allowAutoRepair && backupFailed {
        warnings.append("backup_failed: ingress kept runtime guard active without mutation")
    }

    let shouldPersist = allowMutation && repairedAppData != nil

    return AppDataIngressResult(
        source: source,
        operationId: operationId,
        cleanView: cleanView,
        repairedAppData: repairedAppData,
        repairSummary: summary,
        orchestratorResult: orchestratorResult,
        shouldPersist: shouldPersist,
        passiveStatus: passiveStatus,
        warnings: warnings,
        triggeredOrchestrator: triggeredOrchestrator,
        appDataHashBefore: appDataHashBefore,
        appDataHashAfter: computeAppDataHash(workingAppData)
    )
}

// MARK: - Internals

private func composePassiveStatus(
    appliedCount: Int,
    pendingRepairs: Int,
    auditOnly: Int,
    backupFailed: Bool
) -> AppDataIngressPassiveStatus {
    if backupFailed {
        return AppDataIngressPassiveStatus(line: "数据正在自动整理，稍后同步", tone: .backupFailed)
    }
    if appliedCount > 0 {
        return AppDataIngressPassiveStatus(line: "已自动修复 \(appliedCount) 个旧版本问题", tone: .autoRepaired)
    }
    if pendingRepairs > 0 {
        return AppDataIngressPassiveStatus(line: "\(pendingRepairs) 个待自动修复", tone: .busy)
    }
    if auditOnly > 0 {
        return AppDataIngressPassiveStatus(line: "\(auditOnly) 个已隔离，不影响训练建议", tone: .auditPending)
    }
    return AppDataIngressPassiveStatus(line: "数据已自动检查", tone: .ok)
}

private func generateOperationId(
    source: AppDataIngressSource,
    accountId: String?,
    hash: String,
    clock: RuntimeGuardClock
) -> String {
    let scope = accountId ?? "unscoped"
    let hashSuffix = String(hash.suffix(8))
    let ms = Int(clock.now().timeIntervalSince1970 * 1000.0)
    let msRadix36 = String(ms, radix: 36)
    let rand = Int.random(in: 0..<1_000_000)
    let randRadix36 = String(rand, radix: 36)
    return "ingress_\(source.rawValue)_\(scope)_\(hashSuffix)_\(msRadix36)_\(randRadix36)"
}
